File: backend/app/models/customer.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, Float, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Customer(Base):
    __tablename__ = "customers"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    company_name: Mapped[str] = mapped_column(String(255), nullable=False)
    contact_person: Mapped[str | None] = mapped_column(String(255), nullable=True)
    email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    gst_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    credit_limit: Mapped[float] = mapped_column(Float, default=0)
    credit_used: Mapped[float] = mapped_column(Float, default=0)
    price_tier: Mapped[str] = mapped_column(String(50), default="standard")
    owner_id: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc)
    )

    carts: Mapped[list["CustomerCart"]] = relationship("CustomerCart", back_populates="customer", cascade="all, delete-orphan")

File: frontend/src/types/customer.ts
export interface Customer {
  id: number
  company_name: string | null
  contact_person: string | null
  email: string
  phone: string
  gst_number: string
  credit_limit: number
  credit_used: number
  price_tier: string
  owner_id: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface CustomerFormData {
  company_name: string | null | undefined
  contact_person: string
  email: string
  phone: string
  gst_number: string
  credit_limit: number
  price_tier: string
}
